### PR TITLE
[Impeller] Add support for specialization constants redux.

### PR DESCRIPTION
**Pull request by jonahwilliams**
Reland of https://github.com/flutter/engine/pull/47432

Also includes:
  * https://github.com/flutter/engine/pull/47617
  * https://github.com/flutter/engine/pull/47637
  
Fixes the performance on iOS by removing blocking on compilation of shaders. From local testing this has identical before/after numbers. Additional, ensures that we don't unecessarily specialize vertex shaders and notes this restriction in the documentation.

----

Adds support for Specialization constants to Impeller for our usage in the engine. A motivating example has been added in the impeller markdown docs.

Fixes https://github.com/flutter/flutter/issues/136210
Fixes https://github.com/flutter/flutter/issues/119357

### CODE REVIEW

File: impeller/renderer/backend/metal/pipeline_library_mtl.mm
Comment (jonahwilliams) [inline impeller/renderer/backend/metal/pipeline_library_mtl.mm:55]: This is still a bit iffy, but we improve performance by not blocking on the specialization, it all goes into the same pipeline future which is only blocking once we actually trry to use it.

Since the compilation is out of process anyway, blocking on it is pointless and only serves to park the thread.

File: impeller/renderer/backend/metal/pipeline_library_mtl.mm
Comment (jonahwilliams) [inline impeller/renderer/backend/metal/pipeline_library_mtl.mm:137]: Todo: add back

Comment (jonahwilliams) [reply impeller/renderer/backend/metal/pipeline_library_mtl.mm:137]: Actually, i see this issue is closed. Is this intended to be removed now?

File: impeller/renderer/backend/metal/pipeline_library_mtl.mm
Comment (bdero) [reply impeller/renderer/backend/metal/pipeline_library_mtl.mm:59]: Feels like we're writing javascript here with this chain of nested completion callbacks. ;) Looks safe to me, though.

**Review by bdero — APPROVED** [review]
LGTM modulo latch question. Crasher fix & new changes make sense.

File: impeller/renderer/backend/metal/pipeline_library_mtl.mm
Comment (jonahwilliams) [reply impeller/renderer/backend/metal/pipeline_library_mtl.mm]: Removed. This was left over from an earlier iteration.

**Pull request merged**
